### DISCUSSION TITLE
Dense QCQP in python

### DISCUSSION THREAD

**Discussion by dlcfort** [giaf/hpipm#139] [Q&A] [answered]
Does anyone have an example implementing a dense QCQP problem in python? Will I have to create my own wrapper to implement this?

**Comment by giaf** [comment]
At the moment there is not a python wrapper for the dense QCQP solver.
It should be rather straightforward to add it similarly to here https://github.com/giaf/hpipm/pull/135
Or if there is interest I can put it in my todo list.

**Reply by dlcfort** [reply]
Thanks, I think I will work on implementing a wrapper for dense QCQP in python. However, is it possible to create multiple quadratic inequality constraints for the dense_qcqp? The paper "Introducing Quadratic Programming Framework in HPIPM" suggests that we can set multiple Hk matrices where k is in the set Hq = {1, ... nq}. If this is possible, is there an example where this is done? I only see One Hq matrix being set in x_dense_qcqp.c. Unless I'm missing something. 

Thanks

**Reply by dlcfort** [reply]
After looking at DENSE_QCQP_SET_HQ() function closer, it looks like we just have to create the HQ pointer input as a stacked matrix each containing each Hk submatrix. Is that correct?

**Reply by giaf** [reply]
Yes the number of quadratic inequality constraints is `nq`, and each inequality constraint matrix has size `nv*nv` and is stored in column major.
So in a way, you need to pass an array of length `nq*nv*nv` where the first `nv*nv` elements correspond to the first matrix stored in column major, the following `nv*nv` elements to the second matrix stored in column major, and so on.

This is similar to the linear term of the quadratic inequality constraints too, that is passed as an array of length `nq*nv`, where the first `nv` elements correspond to the first vector, the following `nv` elements to the second vector, and so on.

**Reply by dlcfort** [reply, marked as answer]
Great, I just finished created a forked repo with an implemented wrapper for the dense qcqp. See link below.

https://github.com/dlcfort/hpipm
